File: moqt/fetch.go
package moqt

import (
	"context"
	"fmt"
)

type FetchRequest struct {
	BroadcastPath BroadcastPath
	TrackName     TrackName
	Priority      TrackPriority
	GroupSequence GroupSequence

	ctx context.Context
}

// Context returns the request's context. To change the context, use
// [FetchRequest.WithContext].
//
// The returned context is always non-nil; it defaults to the
// background context.
func (r *FetchRequest) Context() context.Context {
	if r.ctx != nil {
		return r.ctx
	}
	return context.Background()
}

// WithContext returns a shallow copy of r with its context changed
// to ctx. The provided ctx must be non-nil.
func (r *FetchRequest) WithContext(ctx context.Context) *FetchRequest {
	if ctx == nil {
		panic("nil context")
	}
	r2 := new(FetchRequest)
	*r2 = *r
	r2.ctx = ctx
	return r2
}

// Clone returns a deep copy of r with its context changed to ctx.
// The provided ctx must be non-nil.
func (r *FetchRequest) Clone(ctx context.Context) *FetchRequest {
	if ctx == nil {
		panic("nil context")
	}
	r2 := new(FetchRequest)
	*r2 = *r
	r2.ctx = ctx
	return r2
}

type FetchHandler interface {
	ServeFetch(w *GroupWriter, r *FetchRequest)
}

type FetchHandlerFunc func(w *GroupWriter, r *FetchRequest)

func (f FetchHandlerFunc) ServeFetch(w *GroupWriter, r *FetchRequest) {
	f(w, r)
}

func validateFetchHandler(handler FetchHandler) error {
	if handler == nil {
		return fmt.Errorf("moqt: fetch handler cannot be nil")
	}
	if f, ok := handler.(FetchHandlerFunc); ok && f == nil {
		return fmt.Errorf("moqt: fetch handler function cannot be nil")
	}
	return nil
}

func safeServeFetch(handler FetchHandler, w *GroupWriter, r *FetchRequest) (err error) {
	if e := validateFetchHandler(handler); e != nil {
		return e
	}

	defer func() {
		if p := recover(); p != nil {
			err = fmt.Errorf("panic during fetch handling: %v", p)
		}
	}()
	handler.ServeFetch(w, r)
	return nil
}

File: moq-web/src/fetch.ts
import type { BroadcastPath } from "./broadcast_path.ts";
import type { GroupSequence, TrackName, TrackPriority } from "./alias.ts";
import type { GroupWriter } from "./group_stream.ts";

export interface FetchRequestInit {
	broadcastPath: BroadcastPath;
	trackName: TrackName;
	priority: TrackPriority;
	groupSequence: GroupSequence;
	done?: Promise<void>;
}

export class FetchRequest {
	readonly broadcastPath: BroadcastPath;
	readonly trackName: TrackName;
	readonly priority: TrackPriority;
	readonly groupSequence: GroupSequence;
	#done: Promise<void>;

	constructor(init: FetchRequestInit) {
		this.broadcastPath = init.broadcastPath;
		this.trackName = init.trackName;
		this.priority = init.priority;
		this.groupSequence = init.groupSequence;
		this.#done = init.done ?? new Promise(() => {});
	}

	done(): Promise<void> {
		return this.#done;
	}

	withDone(done: Promise<void>): FetchRequest {
		return new FetchRequest({
			broadcastPath: this.broadcastPath,
			trackName: this.trackName,
			priority: this.priority,
			groupSequence: this.groupSequence,
			done,
		});
	}

	clone(done: Promise<void>): FetchRequest {
		return new FetchRequest({
			broadcastPath: this.broadcastPath,
			trackName: this.trackName,
			priority: this.priority,
			groupSequence: this.groupSequence,
			done,
		});
	}
}

export interface FetchHandler {
	serveFetch(w: GroupWriter, r: FetchRequest): void | Promise<void>;
}
